### PR TITLE
test(adapters): probe psmux workaround premises, not just our reaction

### LINKED ISSUE #488: psmux live gate pins no divergence PREMISE, only our reaction to it — CI cannot signal that a workaround became droppable
## Summary

`tests/test_psmux_backend.py` (1648 lines) pins how every psmux workaround is *composed*. Nothing
anywhere observes the upstream behavior each workaround exists **because of**. The live gate
(`tests/test_psmux_live.py`) exercises `new_session` / `new_parked_window` / options / prune / kill —
none of the divergences.

Consequence: when an upstream release fixes a divergence, no test changes colour. The signal
"this workaround is now dead code" and the opposite signal "an assumption we rely on just became
false" both have to come from a human re-reading the upstream changelog. That is exactly the manual
step that stalled #222 for two releases.

Concretely unobserved premises, each currently load-bearing:

| premise | workaround it justifies |
|---|---|
| `pipe-pane` strips dash-flag tokens | sidecar `.ps1` sink (`psmux_backend.py:923-965`) |
| `select-window -t sess:@N` is rejected | `select_window` override + `_window_index` (`:765-817`) |
| `switch-client` reports dispatch, not effect | `#{session_attached}` delta in `_client_left` (`:844-921`) |
| `kill-window` on a bad target exits 0 | `check=False` swallow (`tmux_base.py:337`) |
| `set -w` / `show -wqv` have no per-window storage | the whole `@opt_@N` channel (`:340-638`) |
| `has-session -t =name` resolves the `=` form | the `new_session` belt (`:213-217`) |
| `psmux -V` leads with `tmux X.Y.Z` | the version gate (`:992`) |
| value shapes that survive a `set-option` round trip | `_transportable` (`:457-495`) |

## Ask

Extend the live gate with a premise-probe block: same zero-token contract (parked
`pwsh -NoProfile -Command exit 0` windows, no coding CLI ever launched), asserting the *observed*
upstream behavior rather than our reaction to it. A probe that starts failing is the intended
outcome — it means the workaround above it can go.

Two things make this cheap now:

- `PSMUX_DATA_DIR` (upstream `5152a2c`, ships in 3.3.8) lets the suite point at a throwaway registry
  root instead of `~/.psmux`, so premise probes cannot interact with a developer's live sessions or
  with the reaper.
- The probes are the same commands #222's admission checklist runs by hand — writing them once
  replaces the checklist with a gate.

## Also in scope

Apply the ablation rule to the negative assertions in `tests/test_psmux_backend.py` — delete the
gating code and confirm each test FAILS before trusting it. Candidates: `:96`, `:114`, `:275`, `:414`,
`:824`, `:957`, `:1618-1647`. A "X is refused" assertion passes for every reason a value can be
absent, and several of these guard exactly the workarounds above.

Surfaced by the pre-3.3.8 bughunt while auditing what would catch a regression if the psmux/psmux#482
and #497 workarounds were dropped on the 3.3.8 admission — the answer is nothing

Refs: #222

### PR DESCRIPTION
**Pull request by dracic**
Closes #488

## What

Eight **premise probes** in the Windows live gate. Each drives raw `psmux` argv against a throwaway
session and asserts the *observed upstream behavior* that a workaround in `psmux_backend` exists
because of — never the backend verb the workaround implements.

The failure semantics are inverted on purpose: **a red probe is the intended signal**, and its
message names the workaround that has become droppable. That is the manual step from #222's
admission checklist turned into an assertion.

Plus the ablation half: every negative assertion the issue listed was ablated for real — guard
deleted, test run, result recorded, guard restored.

## Why it was needed, concretely

Four of the eight premises had **already flipped** on psmux `main` (`d56d777`, `6c76ff9`,
`33ea109`, `8edd1cb`/`194fac0`) and nothing in this repo changed colour.

## Every premise was observed before it was asserted

All eight were driven against the installed 3.3.7 build first. Two are sharper than the issue
described:

- `set-option -w` does not merely lack per-window storage — it **silently lands at session scope**
  (`show-options -wqv` reads empty while the same key at session scope reads the value back).
- `kill-window` on an unresolvable target exits **0 and destroys the ACTIVE window**. The probe pins
  exactly that: the session survives and the vanished window is precisely the one that was active.

## Upstream

Tracing the `=`-form kill divergence to its cause produced **psmux/psmux#558**: `parse_target`
strips the `=` for routing but `PSMUX_TARGET_FULL` keeps the raw argument, so the server's
`kill-session` fallback compares `"name" == "=name"` and never kills. Verified identical at `v3.3.7`
and `main`.

A second candidate (CLI value corruption for `'` and non-ASCII whitespace) was traced and
**deliberately not filed** — `e41886f` already fixed it on `main` via a Unicode-aware
`quote_arg_if_needed`. Consequence worth flagging for the 3.3.8 admission: those two `_transportable`
branches become droppable, and the round-trip probe here will go red and say so.

## Ablation results

| verdict | assertions |
|---|---|
| bite on their own assertion | `:276`, `:404`, `:957` |
| belts — a neighbouring `==`/`raises` fires first | `:96`, `:114`, `:824`, `:1618` |
| **vacuous, repaired** | `:1625` — `last_fallback` defaulted false, so the `-l` leg was unreachable regardless of the early-return guard it claimed to cover |
| **wrong-reason failure, repaired** | `:1647` — reached its assertion only via `IndexError` from a fixture that ran dry |

The issue's list also mislabels `:414`: it is a *positive* assertion. The negative one in that test
is `:404`.

Five further weakly-discriminating negatives outside the issue's scope (they ablate into a
`TypeError` rather than reaching their assertion) were recorded in the deferred-work ledger rather
than pulled into this change.

## Invariants held

- **Zero tokens.** Parked windows run `pwsh -NoProfile -Command exit 0`; no coding CLI is launched.
- **No developer session is ever reachable.** Every probe gets its own throwaway session with a
  verified teardown. `PSMUX_DATA_DIR` isolation is applied only when the installed build
  demonstrably honors it. Bare `switch-client -l` is not probed at all — it has no target form, so
  under `$TMUX` it would move the developer's own client.
- **No workaround dropped, no `_LAST_UNSUPPORTED` bump.** This builds the instrument; #222 owns the
  admission.
- **CI untouched** — the live gate stays a manual gate.

## Verification

- `tests/test_psmux_live.py` + `tests/test_psmux_backend.py`: 154 passed on Windows with psmux 3.3.7
- full suite: no psmux-related failures
- `trunk check`: no issues
- the version probe's reachability was proven by inverting its regex and observing the failure —
  as first written it was **vacuous**, since `available()` applies the same regex and skips the test
  on the very failure the probe exists to catch

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved fallback switching so successful client moves no longer trigger unnecessary fallback behavior.
  * Corrected handling for scripted attachment and switching scenarios.

* **Tests**
  * Expanded live checks for session targeting, window selection, option handling, client commands, and pipe-pane behavior.
  * Added coverage for unresolved targets and option-value transport edge cases.

* **Documentation**
  * Documented verified compatibility with supported `psmux` behavior and recorded updated validation details in the changelog.

### CODE REVIEW

**Comment by pbean** [comment]
Reviewed and pushed maintainer-side fixes as `4045e0e` — all in `tests/test_psmux_live.py`; `psmux_backend.py`, `test_psmux_backend.py` and the CHANGELOG entry are untouched (the entry stays accurate). Nice work: the inverted failure semantics and the "an unobserved premise must fail, never pass" guard in the pipe-pane probe are the right shape, and the two premises that turned out sharper than assumed (`-w` landing at session scope, the `=`-form kill traced to psmux/psmux#558) are exactly what this gate is for.

### Changes

1. **Spaced tmp path skips the pipe-pane probe** (`test_premise_pipe_pane_strips_dash_flag_tokens`). Added `" "` to the ``("$", "`")`` refusal tuple and extended the skip reason. psmux's quote re-parse splits a spaced path — a `C:\Users\John Doe` profile is enough — so the sidecar never spawns, and the probe would red *permanently* while its message blamed the transient psmux/psmux#482 spawn race. `pipe_pane`'s own docstring already lists a spaced path as an untested ceiling, so skipping is the honest reading; a red there is not.

2. **`psmux_data_root` now has a leak-verified teardown**, mirroring the `probe` fixture's. This is the one session in the module that can land in a developer's real registry — a build that ignores `PSMUX_DATA_DIR` is precisely the branch the fixture exists to detect — and it had only a bare best-effort kill. The check runs in **both** views (the isolated `env` and the default registry) and fails loud with the session name and "kill it manually" on a leak or on any probe error.

3. **`probe setup: ` prefix on the scaffolding assertions**, so an instrument-failure red is distinguishable from a premise red (the module docstring promises each message names the droppable workaround, and these name none). Applied at: `_raw_new_session` (both), `_mint_probe_window`, `_active_window` (both), both fixture teardowns, the select-window starting-focus and both index-resolution asserts, the attached-count asserts, the roundtrip show/unset asserts, the pipe-pane attach/trigger and "sidecar never ran" asserts, and the kill probe's parked-windows precondition. The premise assertions keep their bare form. One docstring sentence states the convention.

4. **`switch-client` probe runs with `$TMUX`/`$TMUX_PANE` scrubbed** (and its `#{session_attached}` read, for consistency). The premise is "no client to move", and an inherited `$TMUX` says otherwise: when pytest itself runs inside psmux there *is* a client — yours. At 3.3.7 the verb is inert so nothing moves, target routing is port-file based, and rc is dispatch-only `0`, so the probe stays green today either way. But on the effect-reporting build this probe exists to catch, the switch would drag your client into the throwaway session and stay green *through the very premise flip*. A small local scrub rather than `_new_session_env` — that helper strips `CLAUDE_CODE_*` and sets `PSMUX_ALLOW_NESTING`, neither of which this call needs — and it copies `os.environ`, so the fixture's `PSMUX_DATA_DIR` still routes both verbs to the registry the probe session lives in.

5. **`probe` teardown also catches `subprocess.TimeoutExpired`.** `_plain_has_session` goes through raw `_run`, and `_run` propagates a timeout even under `check=False` (its docstring says so), so today a hung psmux raises *out of* the teardown instead of reporting the leak.

6. Two micro-fixes: `_mint_probe_window` asserts the derived id against `[^:]+:@\d+` before returning — the same qualified-id guard the prune test applies to its mints, so a degraded mint fails at the mint instead of surfacing frames later as a bare `rsplit` IndexError; and the kill probe's "therefore last" comment now states the real safety property (every probe owns its own session) rather than a file order nothing enforces under xdist — it stays last for reader flow only.

### Non-blocking, accepted as-is

- **F6** — `_pwsh_quote` is a private import and the only new POSIX coupling this PR adds: a rename upstream in `psmux_backend` would break *collection* on POSIX, where the module is skipped and nobody would see it coming. Accepted: the alternative is duplicating the quoting rule in the test, which is worse — a divergent copy would silently stop matching what `pipe_pane` actually ships.
- **F7** — the default-`last_fallback` True-path coverage gap you found by ablation is subsumed by what landed; no separate test needed.

### Two small notes on the brief, for your call

- The client-verbs probe has one `display-message` read, not two, so the scrub covers that call plus `switch-client`.
- Three asserts in `test_premise_window_scoped_option_write_lands_at_session_scope` are arguably instrument-failure sites too (`set-option -w no longer even accepts a window target`, and the two read-rc checks at `-w` and session scope) but were not in the prefix list, so I left them bare — they read as premise preconditions there. Say the word if you want them prefixed.

### Verification

`uv run pytest -q` green (4752 passed, 38 skipped, 5 xfailed) with all 9 live tests **skipped** — `sys.platform == "win32"` gates the module, so nothing here was executed against a real psmux on my side. `uv run pyright` clean; `trunk check --all --no-fix` clean across 246 files (trunk, not `ruff format`).

I did ablate the new leak guard rather than trust it: with a fake `psmux` whose `has-session` always exits 0, the fixture errors with `probe setup: data-probe session bmad-loop-data-probe-… survived teardown; kill it manually`; with a fake whose `has-session` exits 1, that assert stays silent and the next guard down the chain fires instead. The temporary platform-gate edit that made those runs possible was reverted before the commit (file byte-identical to pre-ablation).

### Ask

Could you re-run the live gate on your psmux 3.3.7 rig? Specifically: **confirm the env-scrubbed `switch-client` probe stays green** — that is the one change with a live behavioral surface, and the prediction is that stripping `$TMUX` changes nothing at 3.3.7 because the verb is inert and routing is port-file based. If it reds, the premise is more interesting than the fix. Also worth one run from a session started *inside* psmux, since that is the case the scrub is for. Holding off on `@codex review` and on merge until then.

File: tests/test_psmux_live.py [hunk 1]
```diff
@@ -99,3 +112,393 @@ def test_prune_kills_only_the_owning_projects_window(tmp_path: Path, monkeypatch
...
+        # a build ignoring PSMUX_DATA_DIR is the very branch this fixture exists
+        # to detect — so verify the kill in BOTH views rather than trusting a
+        # best-effort call whose target may not be the registry it created in.
+        try:
+            leaked = _plain_has_session(mux, session, env=env) or _plain_has_session(mux, session)
+        except (OSError, TmuxError, subprocess.TimeoutExpired):
+            leaked = True
+        assert (
+            not leaked
+        ), f"probe setup: data-probe session {session} survived teardown; kill it manually"
```
[the remark this answers is not in the document]
Comment (pbean) [reply tests/test_psmux_live.py:222]: Half valid, and the valid half is a good catch — fixed in `8d984e9`.

**`psmux_data_root` (207-217): agreed, applied.** That fixture cannot use `kill_session` because it needs `env=` for `PSMUX_DATA_DIR`, so it goes through raw `_run` — which propagates a timeout even under `check=False`. The kill is now inside the `try`. Observed both ways against a fake psmux whose `kill-session` outlasts the 30s `_run` timeout:

- before: `subprocess.TimeoutExpired: Command '['psmux', 'kill-session', '-t', 'bmad-loop-data-probe-…']' timed out after 30 seconds` — no session name, no instruction
- after: `probe setup: data-probe session bmad-loop-data-probe-… survived teardown; kill it manually`

A kill that hung is exactly when the session is most likely still standing, so it has to reach the report rather than escape past it.

**`probe` fixture (237-244): declining, the concern does not reach it.** That one calls `mux.kill_session(session)`, not raw `_run`, and `PsmuxMultiplexer.kill_session` (`psmux_backend.py:219-227`) already wraps its `_run` in `except (subprocess.SubprocessError, OSError)`. `subprocess.TimeoutExpired` is a subclass of `SubprocessError`, so a hung kill is swallowed there and never reaches the `finally`; it also early-returns when the binary is missing, so `OSError` cannot escape either. Moving that call inside the `try` would be a no-op. The asymmetry is now recorded in a comment beside the catch so it does not read as an oversight.

**Comment by pbean** [comment]
### Bot round on `4045e0e` — addressed in `8d984e9`

**CodeRabbit** (1 actionable, Minor): the teardown kill in `psmux_data_root` ran *outside* the `try`, so a `TimeoutExpired` from the kill itself escaped before the new leak check could run. **Valid — applied.** It is the same defect the leak guard exists to close, one line earlier: that fixture needs `env=` for `PSMUX_DATA_DIR`, so it must use raw `_run`, and raw `_run` propagates a timeout even under `check=False`. Observed both ways against a fake psmux whose `kill-session` outlasts the 30s `_run` timeout:

- before — `subprocess.TimeoutExpired: Command '['psmux', 'kill-session', '-t', 'bmad-loop-data-probe-…']' timed out after 30 seconds` (no session name, no instruction)
- after — `probe setup: data-probe session bmad-loop-data-probe-… survived teardown; kill it manually`

The same comment said it "also applies to" the `probe` fixture's `mux.kill_session(session)`. **Declined:** `PsmuxMultiplexer.kill_session` (`psmux_backend.py:219-227`) already catches `(subprocess.SubprocessError, OSError)` and `TimeoutExpired` is a `SubprocessError` subclass, so nothing escapes there — the move would be a no-op. Recorded as a comment beside the catch so it does not read as an oversight next time. Reasoning is in the review thread.

**Codex** auto-reviewed `4045e0e` and found no major issues (this repo has Codex set to auto-review, so it ran without my asking — I had said I would hold off). **Greptile** is inert on this PR: "Your trial has ended", which is a permanent non-answer, not a pass.

### Correction to my verification claim above

⚠️ **GitHub CI has not run on either of my commits.** The green CI on this PR is from `8e9ed10` only — there is no workflow run for `4045e0e` or `8d984e9`, and no `action_required` run queued for this branch either (other branches in this repo do have runs awaiting approval, so the gating is real). `ci.yml` has no `workflow_dispatch`, so I cannot trigger one. My "clean" claim rests on local runs: `uv run pytest -q -n auto` (4752 passed, 38 skipped, 5 xfailed, live module skipped), `uv run pyright` (0 errors), `trunk check --all --no-fix` (246 files clean) — re-run in full after `8d984e9`. Someone should get CI to run on the real head before this merges; worth checking the PR page for an "awaiting approval" banner I cannot see through the API.

### Unchanged

Everything in my previous comment still stands, including the two accepted non-blocking notes (F6 `_pwsh_quote` coupling, F7 subsumed coverage) and the two brief discrepancies flagged for your call. The **ask is also unchanged**: @dracic, please re-run the live gate on your psmux 3.3.7 rig and confirm the env-scrubbed `switch-client` probe stays green, ideally including one run from a shell started *inside* psmux. Still not merging, and still holding the manual `@codex review` pass until after your live run.

**Comment by pbean** [comment]
@codex review

**Comment by pbean** [comment]
### Conflict resolved — CI is green on the real head

`3130900` merges `main` into the branch. `CHANGELOG.md` was the only conflict: both sides had prepended an entry to the same position in `### Added` — #488's psmux probes here, #502's `gate:` entry on main. Both kept, newest first, which is the convention each side independently followed. Merge commit rather than a rebase, so @dracic's commit is not rewritten on his fork.

**Correcting my previous comment.** I said CI might be gated on fork-workflow approval. That was wrong. `pull_request` workflows run against `refs/pull/517/merge`, and while the branch conflicted GitHub could not build that ref — so **no run was created at all**, not even an `action_required` one, which is what misled me. The tell I should have checked first was `mergeable: CONFLICTING` / `mergeStateStatus: DIRTY`. After the merge push, `refs/pull/517/merge` resolves and CI fired within seconds:

**All 10 checks SUCCESS on `3130900`** — test py3.11/3.12/3.13/3.14, test windows py3.11/py3.14, `version-sync`, `lint (trunk)`, `typecheck (pyright)`, CodeRabbit. The two Windows jobs matter here: they are the only ones where this module is *collected* rather than skipped at platform level, so they exercise the `_pwsh_quote` import path that F6 flags as this PR's one POSIX coupling.

Read as a trial merge, the merged tree differs from `main` in exactly this PR's four files — nothing else came along. Locally: `uv run pyright` clean, `trunk check --all --no-fix` clean (247 files), suite green apart from `tests/test_module_skills_sync.py`, which reds only because my gitignored `.claude/skills/` fork is stale against the canonical skill #502 updated — that test `pytest.skip`s when those trees are absent, which is the CI condition, and the PR diff touches neither side. Not this PR's, and CI agrees.

### Bot state

- **CodeRabbit** — the one finding is fixed in `8d984e9` and CodeRabbit has confirmed both halves, including re-reading `kill_session` itself to verify that the `probe` fixture needed no change. Thread marked addressed; no open findings.
- **Codex** — clean, but ⚠️ on `8d984e9`: the review was requested at 20:05 and the merge landed ~20:11, so **the current head `3130900` has not been through Codex.** The only thing it adds is the CHANGELOG resolution, but the gate is "zero comments on the head that merges", so one more `@codex review` now would close it properly.
- **Greptile** — inert on this PR ("Your trial has ended"): a non-answer, not a pass.

### Unchanged

@dracic, the ask still stands: re-run the live gate on your psmux 3.3.7 rig and confirm the env-scrubbed `switch-client` probe stays green, ideally including one run from a shell started *inside* psmux, since that is the case the scrub exists for. Not merging.

**Comment by dracic** [comment]
@pbean live gate re-run complete on the Windows psmux rig.

- Installed build: `tmux 3.3.7` / `psmux 3.3.7 (05cc5d4 2026-07-20)`.
- Outside psmux (`TMUX` unset): `9 passed in 143.78s`.
- From a throwaway shell started inside psmux: `9 passed in 64.31s`. This includes `test_premise_client_verbs_exit_zero_with_no_client_to_move`, so the env-scrubbed `switch-client` probe stayed green in the inherited-`$TMUX` case.
- Teardown checks found no surviving `bmad-loop-test-*`, `bmad-loop-data-probe-*`, runner, or debug session.
- Zero-token contract held: only the live-test parked PowerShell windows ran; no coding CLI was launched.

One wrapper-only false alarm was diagnosed before reporting: I initially used unquoted `@gate_*` option names in an outer PowerShell polling harness, so PowerShell treated them as splatting expressions and the marker was never written/read. A separate throwaway debug session with quoted option names reported `TMUX_PRESENT=True`, `SET_EXIT=0`, and `quoted_marker=inside`, then tore down cleanly. The pytest pane capture independently showed the successful 9/9 inside-psmux run; this was not a repo or live-gate failure.

No code changes were needed.
